[Enhancement] documenting the images variables, `reference` and `referenceWithTag` (#1101)
kyverno/website: content/en/docs/Writing policies/variables.md

      "registry": "https://ghcr.io",
      "path": "tomcat",
      "name": "tomcat",
      "tag": "9",
      "reference":"https://ghcr.io/tomcat:9",
      "referenceWithTag":"https://ghcr.io/tomcat:9"
    }
  },
  "initContainers": {
    "vault": {
      "registry": "https://ghcr.io",
      "path": "vault",
      "name": "vault",
      "tag": "v3",
      "reference":"https://ghcr.io/vault:v3",
      "referenceWithTag":"https://ghcr.io/vault:v3"
    }
  },
  "containers": {
    "nginx": {
      "registry": "https://docker.io",
      "path": "library/nginx",
      "name": "nginx",
      "digest": "sha256:5f44022eab9198d75939d9eaa5341bc077eca16fa51d4ef32d33f1bd4c8cbe7d",
      "reference": "https://docker.io/library/nginx@sha256:5f44022eab9198d75939d9eaa5341bc077eca16fa51d4ef32d33f1bd4c8cbe7d",
      "referenceWithTag": "https://docker.io/library/nginx:"

>     }

Reviewer: Please replace the tomcat example above with this one so there is one representative example within `containers` and `initContainers`.

`{{images.containers.tomcat.digest}}`

6. Reference the readable reference for the image (along with the image digest)

> `{{images.containers.tomcat.reference}}`

Reviewer: What's shown above does not include the image digest so I'm confused.

Author: `reference` is like "registry/path@digest" but since we rather have tag in our example (they are mutually exclusive), `reference` becomes "registry/path:tag". hence, `reference` is the readable reference for the image (along with the image digest), but when tag is the identifier, rather than digest we use tag.

Reviewer: Ok, so the description here shouldn't say "along with the image digest" because that's only in the case where the image was originally specified with a digest and not tag, correct?

Author: I mean here I chose a more generic definition. So yes, if we want to be specific, we should change that phrase.

7. Reference the readable reference for the image (along with the image tag)

`{{images.containers.tomcat.referenceWithTag}}`

> For container `tomcat`, `referenceWithTag` corresponds to `https://ghcr.io/tomcat:9`. Since the image is specified with a `tag` and not a `digest`, `reference` variable will also hold the same value. `reference` enables referencing the image using the image `digest` and when the image is specified with a `tag`,  `reference` enables referencing it using the image `tag` similar to `referenceWithTag`.

Reviewer: ```suggestion
For the container `tomcat`, `referenceWithTag` corresponds to `https://ghcr.io/tomcat:9` since the image was specified with a tag and not a digest. The `reference` variable will also hold the same value. `reference` enables referencing the image depending on how it was originally specified: if it was specified with a tag then `reference` will contain the tag; if specified with digest then `reference` will contain the digest.
```

`{{images.containers.tomcat.digest}}`

6. Reference the readable reference for the image (along with the image digest)

`{{images.containers.tomcat.reference}}`

7. Reference the readable reference for the image (along with the image tag)

`{{images.containers.tomcat.referenceWithTag}}`

> For the container `tomcat`, `referenceWithTag` corresponds to `https://ghcr.io/tomcat:9` since the image was specified with a tag and not a digest. The `reference` variable will also hold the same value. `reference` enables referencing the image depending on how it was originally specified: if it was specified with a tag then `reference` will contain the tag; if specified with digest then `reference` will contain the digest.

Reviewer: But if the original image was specified with a digest and not a tag, then `referenceWithTag` would basically always show an empty string? Ex., `"referenceWithTag": "https://docker.io/library/nginx:"`?

Author: yes, `referenceWithTag` is like "registry/path:tag" so if we have have digest and no tag, `referenceWithTag` would be empty string i.e. **registry/path:**

`{{images.initContainers.vault.digest}}`

6. Reference the readable reference for the image (along with the image digest)

> `{{images.initContainers.vault.reference}}`

Reviewer: Ditto above. Confused.

7. Reference the readable reference for the image (along with the image tag)

`{{images.initContainers.vault.referenceWithTag}}`

> For initContainer `vault`, `referenceWithTag` corresponds to `https://ghcr.io/vault:v3`. Since the image is specified with a `tag` and not a `digest`, `reference` variable will also hold the same value.

Reviewer: Similar suggestion to above.

7. Reference the readable reference for the image (along with the image tag)

`{{images.initContainers.vault.referenceWithTag}}`

For initContainer `vault`, `referenceWithTag` corresponds to `https://ghcr.io/vault:v3` since the image was specified with a tag and not a digest. The `reference` variable will also hold the same value.

Reference the image properties of container `nginx`:

1. Reference the registry URL

`{{images.containers.nginx.registry}}`

2. Reference the path to the image

`{{images.containers.nginx.path}}`

3. Reference the image name

`{{images.containers.nginx.name}}`

4. Reference the image tag

`{{images.containers.nginx.tag}}`

5. Reference the digest

`{{images.containers.nginx.digest}}`

6. Reference the readable reference for the image (along with the image digest)

`{{images.containers.nginx.reference}}`

7. Reference the readable reference for the image (along with the image tag)

`{{images.containers.nginx.referenceWithTag}}`

For the container `nginx`, `referenceWithTag` corresponds to `https://docker.io/library/nginx:`. Since the image was specified with a digest, `reference` corresponds to `https://docker.io/library/nginx@sha256:abcdef1234567890`.

> 

Reviewer: Again, this is redundant. Just one for `images.containers.` and one for `images.initContainers`.
